File: impl/golang/internal/health/liveness.go
package health

// LivenessState represents the five-state liveness classification of a peer.
// FR-H18, FR-H19, FR-H20.
type LivenessState string

const (
	// StateUnknown indicates no heartbeat has been received from this peer.
	StateUnknown LivenessState = "UNKNOWN"
	// StateAlive indicates the peer is within its declared heartbeat window.
	StateAlive LivenessState = "ALIVE"
	// StateSuspect indicates the peer has missed its deadline but is within the grace+suspect window.
	StateSuspect LivenessState = "SUSPECT"
	// StateDead indicates the peer has exceeded the grace+suspect window without a new heartbeat.
	StateDead LivenessState = "DEAD"
	// StateOffline indicates the peer has sent a graceful shutdown sentinel (deadline=0).
	StateOffline LivenessState = "OFFLINE"
)

// String returns the string representation of a LivenessState.
func (s LivenessState) String() string {
	return string(s)
}

// LivenessRecord holds the observer-local liveness state for a single peer.
// This is an in-memory record, not persisted on-chain.
// FR-H18.
type LivenessRecord struct {
	// senderAddress is the EVM address (hex) of the peer whose liveness is tracked.
	senderAddress string
	// currentState is the current liveness classification of this peer.
	currentState LivenessState
	// lastDeadline is the nextHeartbeatDeadline from the most recent valid heartbeat.
	lastDeadline uint64
	// lastSequence is the sequence number of the most recent valid heartbeat.
	lastSequence uint64
	// lastConsensusTimestamp is the consensus timestamp of the most recent valid heartbeat.
	lastConsensusTimestamp uint64
}

// SenderAddress returns the EVM address of the peer.
func (r *LivenessRecord) SenderAddress() string { return r.senderAddress }

// CurrentState returns the current liveness state.
func (r *LivenessRecord) CurrentState() LivenessState { return r.currentState }

// LastDeadline returns the last heartbeat deadline.
func (r *LivenessRecord) LastDeadline() uint64 { return r.lastDeadline }

// LastSequence returns the last sequence number.
func (r *LivenessRecord) LastSequence() uint64 { return r.lastSequence }

// LastConsensusTimestamp returns the last consensus timestamp.
func (r *LivenessRecord) LastConsensusTimestamp() uint64 { return r.lastConsensusTimestamp }

// EvaluateLiveness is a pure function that determines liveness state from a record and current time.
// It does not mutate the record. All time arithmetic uses abstract uint64 timestamps
// (consensus-sourced, not local clock).
//
// Decision tree:
//   - nil record          -> UNKNOWN  (no data)
//   - deadline == 0       -> OFFLINE  (graceful shutdown sentinel)
//   - now <= deadline+GP  -> ALIVE    (within grace window)
//   - now <= deadline+GP+S2D -> SUSPECT (grace expired, suspect window)
//   - otherwise           -> DEAD     (all windows expired)
//
// Overflow safety: The additions (deadline + GracePeriod + SuspectToDead) cannot overflow
// in practice because LastDeadline is bounded by ValidateInboundHeartbeat which enforces
// deadline <= consensusTimestamp + MaxDeadlineDelta (86400). A consensusTimestamp near
// math.MaxUint64 would require ~585 billion years of Unix time, which is physically
// unreachable. Should this assumption ever change, add explicit overflow guards here.
//
// FR-H20, SC-H04.
func EvaluateLiveness(record *LivenessRecord, currentTime uint64) LivenessState {
	if record == nil {
		return StateUnknown
	}
	if record.lastDeadline == ShutdownSentinel {
		return StateOffline
	}
	if currentTime <= record.lastDeadline+GracePeriod {
		return StateAlive
	}
	if currentTime <= record.lastDeadline+GracePeriod+SuspectToDead {
		return StateSuspect
	}
	return StateDead
}

File: impl/typescript/src/health/liveness.ts
/**
 * Liveness state machine.
 *
 * Spec reference: 005 spec.md, data-model.md, contracts/health-observer.md
 *   - FR-H18: LivenessRecord tracks per-peer state
 *   - FR-H19: 5-state model (UNKNOWN, ALIVE, SUSPECT, DEAD, OFFLINE)
 *   - FR-H20: State evaluation algorithm
 *   - FR-H21: Recovery invariant -- any valid heartbeat transitions to ALIVE from any state
 *   - FR-H17: Sequence ordering -- only highest sequence number is authoritative
 *
 * All time values (currentTime, deadlines, constants) are in nanoseconds.
 */

import { LivenessState } from './types.js';
import type { LivenessRecord } from './types.js';
import { GRACE_PERIOD, SUSPECT_TO_DEAD } from './constants.js';

/** One second in nanoseconds. */
const NANOS_PER_SECOND = 1_000_000_000n;

/** Grace period in nanoseconds. */
const GRACE_NANOS = GRACE_PERIOD * NANOS_PER_SECOND;

/** Suspect-to-dead window in nanoseconds. */
const S2D_NANOS = SUSPECT_TO_DEAD * NANOS_PER_SECOND;

/**
 * Evaluate the current liveness state for a peer.
 *
 * FR-H20: Pure function of the LivenessRecord and current time.
 *
 * Algorithm:
 *   - If record is null (no heartbeat ever observed): UNKNOWN
 *   - If lastDeadline === 0n (shutdown sentinel): OFFLINE
 *   - If currentTime <= deadline + GRACE_PERIOD: ALIVE
 *   - If currentTime <= deadline + GRACE_PERIOD + SUSPECT_TO_DEAD: SUSPECT
 *   - Else: DEAD
 *
 * @param record - Per-peer liveness record (null if no heartbeat observed)
 * @param currentTime - Current time in nanoseconds
 * @returns Computed LivenessState
 */
export function evaluateLiveness(
  record: LivenessRecord | null,
  currentTime: bigint,
): LivenessState {
  // No heartbeat ever observed
  if (record === null) {
    return LivenessState.UNKNOWN;
  }

  // Shutdown sentinel (FR-H12)
  if (record.lastDeadline === 0n) {
    return LivenessState.OFFLINE;
  }

  // ALIVE: within deadline + grace period
  if (currentTime <= record.lastDeadline + GRACE_NANOS) {
    return LivenessState.ALIVE;
  }

  // SUSPECT: within deadline + grace + suspect-to-dead
  if (currentTime <= record.lastDeadline + GRACE_NANOS + S2D_NANOS) {
    return LivenessState.SUSPECT;
  }

  // DEAD: exceeded all tolerance windows
  return LivenessState.DEAD;
}

/**
 * Update a LivenessRecord when a new valid heartbeat arrives.
 *
 * FR-H17: Only the highest sequence number is authoritative.
 * FR-H21: Recovery invariant -- any valid heartbeat from any state transitions to ALIVE
 *         (or OFFLINE if deadline === 0n).
 *
 * @param record - Existing record (null for first heartbeat from this peer)
 * @param deadline - nextHeartbeatDeadline from the heartbeat payload (nanoseconds)
 * @param consensusTimestamp - Consensus timestamp from MessageDelivery (nanoseconds)
 * @param sequenceNumber - Sequence number from TopicMessage envelope
 * @param senderAddress - EVM address of the sender
 * @returns Updated LivenessRecord
 */
export function updateLivenessRecord(
  record: LivenessRecord | null,
  deadline: bigint,
  consensusTimestamp: bigint,
  sequenceNumber: bigint,
  senderAddress: string,
): LivenessRecord {
  // FR-H17: Ignore if sequenceNumber <= record.lastSequence (out-of-order or duplicate)
  if (record !== null && sequenceNumber <= record.lastSequence) {
    return record;
  }

  // Determine new state based on deadline
  const newState = deadline === 0n ? LivenessState.OFFLINE : LivenessState.ALIVE;

  return {
    senderAddress,
    currentState: newState,
    lastDeadline: deadline,
    lastSequence: sequenceNumber,
    lastConsensusTimestamp: consensusTimestamp,
  };
}
